llama : offload to RPC in addition to other backends

rgerganov opened the pull request:
This patch adds support for offloading layers to RPC servers in addition to other non-RPC backends. For example if you build with `-DLLAMA_CUDA=ON -DLLAMA_RPC=ON`, then you can offload to local GPU and remote server(s):

```bash
$ bin/main -m ../models/ggml-model-f16.gguf -p "Hello, my name is" -n 64 -ngl 99 -s 1236 --rpc localhost:50052
...
ggml_cuda_init: GGML_CUDA_FORCE_MMQ:   no
ggml_cuda_init: CUDA_USE_TENSOR_CORES: yes
ggml_cuda_init: found 1 CUDA devices:
  Device 0: NVIDIA T1200 Laptop GPU, compute capability 7.5, VMM: yes
llm_load_tensors: ggml ctx size =    0,31 MiB
llm_load_tensors: offloading 22 repeating layers to GPU
llm_load_tensors: offloading non-repeating layers to GPU
llm_load_tensors: offloaded 23/23 layers to GPU
llm_load_tensors:        CPU buffer size =   125,00 MiB
llm_load_tensors:      CUDA0 buffer size =  1008,19 MiB
llm_load_tensors:        RPC buffer size =   965,16 MiB
..........................................................................................
```

I have tried to follow the existing patterns in `llama.cpp` and introduced device numbers for RPC servers which always come last.

When copying tensors, we need to handle the case when `src` and `dst` are not on the same backend. For CUDA I had to build with `-DLLAMA_CUDA_NO_PEER_COPY=ON` to make it work.

rgerganov committed:
llama : offload to RPC in addition to other backends

slaren commented:
> When copying tensors, we need to handle the case when `src` and `dst` are not on the same backend. For CUDA I had to build with `-DLLAMA_CUDA_NO_PEER_COPY=ON` to make it work.

The way this is supposed to work is that backends need to check the buffer type of the tensors to determine if they can perform the copy, and return `false` otherwise. The CUDA backend should already do this. What happens exactly if you don't use `LLAMA_CUDA_NO_PEER_COPY`, does it crash somewhere or does it produce wrong results?

slaren commented on ggml-rpc.h:
```diff
@@ -10,12 +10,14 @@ extern "C" {
 
 // backend API
-GGML_API GGML_CALL ggml_backend_t ggml_backend_rpc_init(const char * endpoint);
+GGML_API GGML_CALL void ggml_backend_rpc_setdevice(const char * endpoint, int device);
```
I think it is a bad idea to add this as global state of the RPC backend. For example, this will prevent loading a model with a set of RPC servers, and a different model with a different set of RPC servers at the same time. Is it really necessary? I think it was ok to refer to each RPC server by a endpoint string, rather than a device index.

rgerganov replied:
The intent was to make it consistent with other backends like CUDA and sycl but I agree that having a global state is not great. I will add more special handling in `llama.cpp` and revert the changes in the RPC backend.

rgerganov commented:
> The way this is supposed to work is that backends need to check the buffer type of the tensors to determine if they can perform the copy, and return `false` otherwise. The CUDA backend should already do this. What happens exactly if you don't use `LLAMA_CUDA_NO_PEER_COPY`, does it crash somewhere or does it produce wrong results?

It crashes in `ggml_backend_cuda_buffer_cpy_tensor()` because it assumes that `dst` is allocated by the CUDA backend which is not the case. I think we should add a check in `ggml_backend_tensor_copy()` if both tensors are on the same backend and perform the slow copy if they are not.

slaren committed:
- fix copy_tensor being called on the src buffer instead of the dst buffer

- always initialize views in the view_src buffer

- add RPC backend to Makefile build

- add endpoint to all RPC object names

slaren commented:
https://github.com/rgerganov/llama.cpp/pull/1 should fix it.

slaren committed:
add rpc-server to Makefile

rgerganov commented:
@slaren thanks, I have verified your fix and merged it

slaren commented on llama.cpp:
```diff
@@ -16427,6 +16428,18 @@ struct llama_context * llama_new_context_with_model(
         }
+#endif
+#if defined(GGML_USE_RPC)
+        for (int i = 0; i < (int)model->rpc_servers.size(); i++) {
+            const char * endpoint = model->rpc_servers[i].c_str();
+            ggml_backend_t backend = ggml_backend_rpc_init(endpoint);
+            if (backend == nullptr) {
+                LLAMA_LOG_ERROR("%s: failed to initialize RPC to '%s'\n", __func__, endpoint);
+                llama_free(ctx);
+                return nullptr;
+            }
+            ctx->backends.push_back(backend);
+        }
```
```suggestion
        if (model->n_gpu_layers > 0) {
            for (const auto & endpoint : model->rpc_servers) {
                ggml_backend_t backend = ggml_backend_rpc_init(endpoint);
                if (backend == nullptr) {
                    LLAMA_LOG_ERROR("%s: failed to initialize RPC to '%s'\n", __func__, endpoint);
                    llama_free(ctx);
                    return nullptr;
                }
                ctx->backends.push_back(backend);
            }
        }
```

#7711 made me realize that `ggml_backend_rpc_init` cannot fail because it never attempts a connection. Then it returns a `NULL` buffer type in `get_default_buffer_type`, which is never supposed to happen.

rgerganov replied:
should we also check for `LLAMA_SPLIT_MODE_LAYER` here in addition to `n_gpu_layers > 0`?

slaren replied:
All modes would require initializing the RPC backend. It could be improved to check that the backend is actually used in the model before instantiating it, but mostly the goal here is to avoid issues with `-ngl 0`. I think that the bug on master that causes a RPC backend with a empty name to be initialized always is not present here.

rgerganov committed:
Update llama.cpp

Co-authored-by: slaren <slarengh@gmail.com>

zhouwg commented:
@rgerganov, hello, I'm sorry to disturb you. could you do a more carefully check before your PR merged to master branch? I know you are the brother of the original author of the original author of ggml/llama.cpp/whisper.cpp and you have some privileges. but your modifications in llama.cpp might be have side-effect to other backend developers. Thanks for your understanding.

rgerganov commented:
> could you do a more carefully check before your PR merged to master branch?

My changes have been reviewed and the CI run was successful.

> and you have some privileges and this is the key reason why your PR was approved so quickly although the quality of your PR is might-be need more check.

I don't have any privileges in this project and my code is being reviewed as everybody's else. You'd better have some arguments when saying something about the quality of my changes.

> I'm not sure whether you might did wrong deletions in llama.cpp and cause the troubles for rebase operation by the other community backend developer

When someone is proposing a change it is their responsibility to address review comments and make sure that the change can be applied on current `master`. The changes I have done in `llama.cpp` are trivial and rebasing other changes on top of them is also trivial.

> I hope you can do more carefully check before your PR was merged to master branch.

My change has been reviewed and approved by a core maintainer and the CI run was successful.

zhouwg commented:
> > could you do a more carefully check before your PR merged to master branch?
> 
> My changes have been reviewed and the CI run was successful.
> 
> > and you have some privileges and this is the key reason why your PR was approved so quickly although the quality of your PR is might-be need more check.
> 
> I don't have any privileges in this project and my code is being reviewed as everybody's else. You'd better have some arguments when saying something about the quality of my changes.
> 
> > I'm not sure whether you might did wrong deletions in llama.cpp and cause the troubles for rebase operation by the other community backend developer
> 
> When someone is proposing a change it is their responsibility to address review comments and make sure that the change can be applied on current `master`. The changes I have done in `llama.cpp` are trivial and rebasing other changes on top of them is also trivial.
> 
> > I hope you can do more carefully check before your PR was merged to master branch.
> 
> My change has been reviewed and approved by a core maintainer and the CI run was successful.

Yes, you are right:"When someone is proposing a change it is their responsibility to address review comments and make sure that the change can be applied on current `master`.". my rebase operation has conflict your latest PR in ggml-rpc.cpp.

slaren commented:
@zhouwg I am having a very hard time understating what you are complaining about here. It is the responsibility of PR authors to deal with merge conflicts caused by changes to the code merged to the repository. How could it be otherwise? The RPC backend has been submit to the same quality requirements as any other PRs merged to llama.cpp.

The QNN backend cannot be merged in its current state because it is not a functional backend. It is not correct to say that it has been waiting for review since 4/24/2024 because there is nothing to review until it is a functional backend that people can try. Generally, we will not merge non-functional code in the hope that other people will complete the missing parts.

You have two ways to make it functional: implement the missing operations, or wait until ggml-backend implements support for using the CPU backend for ops not supported by a backend. I understand that you are frustrated, but please try to be patient.

zhouwg commented:
> @zhouwg I am having a very hard time understating what you are complaining about here. It is the responsibility of PR authors to deal with merge conflicts caused by changes to the code merged to the repository. How could it be otherwise? The RPC backend has been submit to the same quality requirements as any other PRs merged to llama.cpp.
> 
> The QNN backend cannot be merged in its current state because it is not a functional backend. It is not correct to say that it has been waiting for review since 4/24/2024 because there is nothing to review until it is a functional backend that people can try. Generally, we will not merge non-functional code in the hope that other people will complete the missing parts.
> 
> You have two ways to make it functional: implement the missing operations, or wait until ggml-backend implements support for using the CPU backend for ops not supported by a backend. I understand that you are frustrated, but please try to be patient.

I'm sorry for that and pls don't care what I said(which is just a personal feeling/thoughts and I still think you are a real AI expert and C/C++ master) in this PR.

1.I provide a general approach based on the existing ggml backend subsystem for CPU&GPU / CPU&NPU mixed inference very easily in PR:[refine backend subsystem for CPU&GPU / CPU&NPU mixed inference more easily for a specified GGML backend](https://github.com/ggerganov/llama.cpp/pull/7641), but you said it's not correct and will not work although it's works very well on my local dev envs.

2.The QNN backend is more mature if we compare to the ggml-rpc backend, this is also a personal feeling.

3.You are the author/maintainer of ggml backend subsystem, so I respect your decision.

airMeng commented on ggml-backend.c:
```diff
@@ -1887,15 +1887,15 @@ ggml_backend_t ggml_backend_sched_get_tensor_backend(ggml_backend_sched_t sched,
-void ggml_backend_view_init(ggml_backend_buffer_t buffer, struct ggml_tensor * tensor) {
+void ggml_backend_view_init(struct ggml_tensor * tensor) {
     GGML_ASSERT(tensor->buffer == NULL);
     GGML_ASSERT(tensor->view_src != NULL);
     GGML_ASSERT(tensor->view_src->buffer != NULL);
     GGML_ASSERT(tensor->view_src->data != NULL);
 
-    tensor->buffer = buffer;
+    tensor->buffer = tensor->view_src->buffer;
     tensor->data = (char *)tensor->view_src->data + tensor->view_offs;
-    ggml_backend_buffer_init_tensor(buffer, tensor);
+    ggml_backend_buffer_init_tensor(tensor->buffer, tensor);
```
This breaks the latest SYCL support. Could I know if all the other backends assumes the buffer already bound to the right tensor?

@slaren @rgerganov for awareness

slaren replied:
It's the same issue that was fixed for the Vulkan backend in #7806, there are more details about why this happens and why the change was necessary there. The best way to fix this for the SYCL backend would be to remove the extras entirely in the same way they were removed from the CUDA backend.

rgerganov replied:
> we are maintaining SYCL during spare time and are still begging for official support from the company

Having a dedicated machine which runs the CI with the SYCL backend would be very helpful

metal3d commented:
I see that you reverted this merge. Does it mean that it will not work with others backend? I'm using Vulkan instead of CUDA (for many reasons), and actually I can see that `rpc-server` binary is linked to libvulkan, but the binary says "create_backend: using CPU backend".

belog2867 commented:
I also encountered such a problem. I compiled the rpc of opencl backend for Android, but it was output when I run it .
 started.create_backend: using CPU backend

But the independent llama-cli can call the opencl backend.

slaren commented:
@zhouwg ok, I have deleted the comment.
